korangar/src/loaders/animation/mod.rs

```diff
@@ -176,95 +183,104 @@ impl AnimationLoader {
+            animations[action_index].frames.iter_mut().for_each(|frame| {
+                frame.size = Vector2::new(size_x, size_y);
+                // Set the point (0, 0) correctly by offseting by max_bottom
+                // The player extra_offset is a small offset to correct the position for
+                // player conjunction with different sizes.
+                frame.offset = Vector2::new(0, -max_bottom - player_extra_offset);
                 for frame_part in frame.frame_parts.iter_mut() {
-                    frame_part.offset.x += offset_width;
-                    frame_part.offset.y += offset_height;
+                    let frame_top_left = frame.offset - (frame.size - Vector2::new(1, 1)) / 2;
+                    let frame_part_top_left = frame_part.offset - (frame_part.size - Vector2::new(1, 1)) / 2;
 
-                    // Precompute the vertex for rendering later.
                     let new_vector = frame_part.size;
-                    let old_origin = frame.offset - (frame.size - frame.remove_offset - Vector2::new(1, 1)) / 2;
-                    let new_origin = frame_part.offset - (frame_part.size - Vector2::new(1, 1)) / 2;
-                    let top_left = new_origin - old_origin;
+                    let top_left = frame_part_top_left - frame_top_left;
                     let bottom_left = top_left + new_vector.y * Vector2::<i32>::unit_y();
                     let bottom_right = top_left + new_vector;
 
                     let texture_top_left = convert_coordinates(top_left, frame.size);
                     let texture_bottom_left = convert_coordinates(bottom_left, frame.size);
                     let texture_bottom_right = convert_coordinates(bottom_right, frame.size);
 
-                    // Scale the vertex (2.-1), (0, -1), (2, 1), (0, 1) to fit texture coordinates
+                    // Scale the vertex (2, -1), (0, -1), (2, 1), (0, 1) to fit texture coordinates
                     // as above.
                     let scale_matrix = Matrix4::from_nonuniform_scale(
                         (texture_bottom_right.x - texture_bottom_left.x) / 2.0,
                         (texture_top_left.y - texture_bottom_left.y) / 2.0,
                         1.0,
                     );
+                    let rotation_matrix = Matrix4::from_angle_z(Rad(-frame_part.angle));
 
                     let new_center = Vector2::new(0.0, (texture_top_left.y - texture_bottom_left.y) / 2.0);
                     let texture_center = (texture_top_left + texture_bottom_right) / 2.0;
                     let translation_matrix = Matrix4::from_translation(texture_center.extend(1.0) - new_center.extend(1.0));
 
-                    frame_part.affine_matrix = translation_matrix * scale_matrix;
+                    frame_part.affine_matrix = translation_matrix
+                        * scale_matrix
+                        * Matrix4::from_translation(Vector3::<f32>::new(0.0, 1.0, 0.0))

```

Comment:
Can you please document why we need to do the two translations? Are they just random stuff that gravity applied?

Reply:
Ok, fixed the documentation of the translation.
May I change the order of translation to make it more safe?

Reply:
I will change a multiplication by a division.

Reply:
Sure! Safer is always better!